In [57]:
from math import asin
from typing import Union
from tqdm import tqdm

import pandas as pd
import numpy as np

from scipy import stats
from statsmodels.stats.meta_analysis import effectsize_smd
from statsmodels.stats import proportion
from statsmodels.stats.power import tt_ind_solve_power
from statsmodels.stats.power import zt_ind_solve_power

In [58]:
import warnings
warnings.filterwarnings('ignore')
warnings.warn('DelftStack')
warnings.warn('Do not show this message')

In [59]:
import matplotlib.pyplot as plt
from google_drive_downloader import GoogleDriveDownloader as gdd

gdd.download_file_from_google_drive(file_id='1IGMgJP_XqQwA6h_w1KYG7TQJhS15h4c4',
                                    dest_path='./gb_sem_8_cw.xlsx');

In [60]:
data = pd.read_excel('gb_sem_8_cw.xlsx')

data.head(10)

USER_ID VARIANT_NAME  REVENUE
0      737      variant      0.0
1     2423      control      0.0
2     9411      control      0.0
3     7311      control      0.0
4     6174      variant      0.0
5     2380      variant      0.0
6     2849      control      0.0
7     9168      control      0.0
8     6205      variant      0.0
9     7548      control      0.0

In [61]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   USER_ID       10000 non-null  int64  
 1   VARIANT_NAME  10000 non-null  object 
 2   REVENUE       10000 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 234.5+ KB


In [62]:
data.shape

(10000, 3)

In [63]:
data.describe()

USER_ID       REVENUE
count  10000.000000  10000.000000
mean    4981.080200      0.099447
std     2890.590115      2.318529
min        2.000000      0.000000
25%     2468.750000      0.000000
50%     4962.000000      0.000000
75%     7511.500000      0.000000
max    10000.000000    196.010000

In [83]:
data.USER_ID.nunique()

6324

In [67]:
data = data.groupby(['USER_ID', 'VARIANT_NAME'], as_index=False).agg({'REVENUE': 'sum'})

In [68]:
data.shape

(7865, 3)

In [69]:
data.groupby('USER_ID', as_index=False).agg({'VARIANT_NAME': 'count'})['VARIANT_NAME'].value_counts()

1    4783
2    1541
Name: VARIANT_NAME, dtype: int64

In [71]:
uniq_ids = (data.groupby('USER_ID', as_index=False).agg({'VARIANT_NAME': 'count'}).query('VARIANT_NAME == 1').USER_ID.values)

In [72]:
data_new = data[data.USER_ID.isin(uniq_ids)].copy(deep=True)

In [73]:
data_new.describe()

USER_ID      REVENUE
count  4783.000000  4783.000000
mean   4994.395777     0.135873
std    2898.618472     3.011392
min       2.000000     0.000000
25%    2476.000000     0.000000
50%    4975.000000     0.000000
75%    7515.000000     0.000000
max    9998.000000   196.010000

In [74]:
df.isna().sum()

USER_ID         0
VARIANT_NAME    0
REVENUE         0
dtype: int64

In [78]:
def continious_result(control: pd.DataFrame,
                      treatment: pd.DataFrame,
                      column: str,
                      n_iters: int = 10_000) -> pd.DataFrame:
    # Статистика по выборкам
    size = control.loc[:, column].shape[0]
    
    control_mean = control.loc[:, column].mean()
    treatment_mean = treatment.loc[:, column].mean()
    
    control_std = control.loc[:, column].std(ddof=1)
    treatment_std = treatment.loc[:, column].std(ddof=1)
    
    # Бутсрап
    booted_diff = []
    for _ in tqdm(range(n_iters)):
        control_sample = control.loc[:, column].sample(n=size, replace=True).values
        treatment_sample = treatment.loc[:, column].sample(n=size, replace=True).values
        booted_diff.append(np.mean(control_sample - treatment_sample))
    
    # Считаем статистику после бустрапа
    md_ci, std_ci = np.mean(booted_diff), np.std(booted_diff, ddof=1)
    left_ci, right_ci = np.percentile(booted_diff, [2.5, 97.5])
    p_value_ci = 2 * (1 - stats.norm.cdf(np.abs(md_ci / std_ci)))
    
    # Считаем мощность эксперимента
    effect_size, _ = effectsize_smd(mean1=treatment_mean, sd1=treatment_std, nobs1=size,
                                    mean2=control_mean, sd2=control_std, nobs2=size)
    power = tt_ind_solve_power(effect_size=effect_size,
                               nobs1=size,
                               alpha=.05,
                               power=None,
                               ratio=1)
    # Формируем отчёт 
    result = pd.DataFrame({'effect_size': effect_size,
                           'alpha': p_value_ci, 
                           'beta': (1-power),
                           'CI': f'[{np.round(left_ci, 3)}, {np.round(right_ci, 3)}]',
                           'difference': md_ci,},
                          index=[column]) 
    return result

In [76]:
control = data_new[data_new['VARIANT_NAME'] == 'control'].copy(deep=True)
treatment = data_new[data_new['VARIANT_NAME'] == 'variant'].copy(deep=True)

In [80]:
control.describe()

USER_ID      REVENUE
count  2390.000000  2390.000000
mean   5020.881590     0.196887
std    2904.850992     4.172201
min       2.000000     0.000000
25%    2517.250000     0.000000
50%    5012.500000     0.000000
75%    7616.000000     0.000000
max    9998.000000   196.010000

In [81]:
treatment.describe()

USER_ID      REVENUE
count  2393.000000  2393.000000
mean   4967.943168     0.074935
std    2892.745368     0.858207
min       4.000000     0.000000
25%    2435.000000     0.000000
50%    4955.000000     0.000000
75%    7379.000000     0.000000
max    9995.000000    23.040000

In [79]:
continious_result(control, treatment, column='REVENUE')

100%|██████████| 10000/10000 [00:02<00:00, 3573.66it/s]


effect_size     alpha      beta               CI  difference
REVENUE    -0.040483  0.158725  0.712143  [-0.004, 0.317]    0.122419

Вывод: Распределение на группы в тестировании сделано некорректно, один и тот же пользователь мог попасть в обе группы и, сделать, либо нет, ключевое действие.

Доверительный интервал [-0.004, 0.317] содержит 0, alpha больше 5% и beta большие 20%, можно говорить об отсутствии статитстически значимых различий.

Нужно сделать правильное распределение на группы. Это исследование я бы не повторял, искал бы другие гипотезы для тестирования.